cmds/exp/disk_unlock/disk_unlock.go

```diff
@@ -162,20 +190,49 @@ func main() {
 
 	verboseLog(fmt.Sprintf("Disk info for %s: %s", *disk, info.String()))
 
+	switch {
+	case !info.SecurityStatus.SecurityEnabled():
+		log.Printf("disk security is not enabled on %v", *disk)
+		return
+	case info.SecurityStatus.SecurityFrozen():
+		// If the disk is frozen, its security state cannot be changed until the next
+		// power on or hardware reset. Disk unlock will fail anyways, so return.
+		// This is unlikely to occur, since someone would need to freeze the drive's
+		// security state between the last AC cycle and this code being run.
+		log.Print("disk security is frozen. Power cycle the machine to unfreeze the disk")
+		return
+	case !info.SecurityStatus.SecurityLocked():
+		log.Print("disk is already unlocked")
+		return
+	case info.SecurityStatus.SecurityCountExpired():
+		// If the security count is expired, this means too many attempts have been
+		// made to unlock the disk. Reset this with an AC cycle or hardware reset
+		// on the disk.
+		log.Print("security count expired on disk. Reset the password counter by power cycling the disk.")
+		return
+	case info.MasterPasswordRev != SkmMPI:
+		log.Printf("disk is locked with unknown master password ID: %X", info.MasterPasswordRev)
+		return
+	}
+
 	// Try using each HSS to unlock the disk - only 1 should work.
 	unlocked := false
+
+TryAllHSS:
 	for i, hss := range hssList {
 		key, err := genPassword(hss, info)
 		if err != nil {
 			log.Printf("Couldn't generate password with HSS %d: %v", i, err)
 			continue
 		}
 
-		if err := sgdisk.Unlock((string)(key), false); err != nil {
-			log.Printf("Couldn't unlock disk with HSS %d: %v", i, err)
-		} else {
-			unlocked = true
-			break
+		for r := 0; r < *retries; r++ {
+			if err := sgdisk.Unlock(string(key), false); err != nil {
+				log.Printf("Couldn't unlock disk with HSS %d: %v", i, err)
+			} else {
+				unlocked = true
+				break TryAllHSS
```

Comment:
what exactly does this do? does it go back to the beginning of the for-loop, or break out of it?

Reply:
it should break out of the loop with the "TryAllHSS" label, which in this case is the outer one